cpu/load_balancer.py

```diff
@@ -0,0 +1,169 @@
+        for package in deps:
+            if not sm.check_installed(package) and not sm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+        smt_op = process.run("ppc64_cpu --smt", shell=True,
+                             ignore_status=True).stderr.decode("utf-8")
+        if "is not SMT capable" in smt_op:
+            self.cancel("Machine is not SMT capable, skipping the test")
+        self.no_threads = self.params.get("no_threads", default=24)
+        self.cpu_cycles = self.params.get("cpu_cycles", default=10000000)
+        self.capacity = self.params.get("capacity", default=100)
+        distro_name = self.detected_distro.name
+        distro_ver = self.detected_distro.version
+        distro_rel = self.detected_distro.release
+        if distro_name == "rhel":
+            if (distro_ver == "7" or
+                    (distro_ver == "8" and distro_rel < "4")):
+                self.cancel("smtstate tool is supported only after RHEL8.4")
+        elif distro_name == "SuSE":
+            if (distro_ver == "12" or (distro_ver == "15" and distro_rel < 3)):
+                self.cancel("smtstate tool is supported only after SLES15 SP3")
+        else:
+            self.cancel("Test case is supported only on RHEL and SLES")
+        process.run("pip3 install pandas", shell=True)
+
+    def run_workload(self, no_threads, cpu_cycles, capacity):
+        """
+        Run the stress-ng workload
+        """
+        cmd = 'nohup stress-ng --cpu %s --cpu-ops %s -l %s &> \
+                /tmp/stress-ng.txt &' % (no_threads, cpu_cycles, capacity)
+        process.run(cmd, shell=True)
+        self.log.info("stress-ng workload started running successfully--!!")
+
+    def mpstat_analyzer(self, on_cpu_count, smt_mode, core):
+        """
+        This function basically capture the mpstat command
+        output for cpu utilization and validate the mpstat data.
+        """
+        cmd = 'mpstat -P ALL 1 1 | awk \'$3 != "all" && $4 > 0\' &> \
+                /tmp/mpstat.log'
+        process.run(cmd, shell=True)
+        with open("/tmp/mpstat.log", "r") as file:
+            lines = file.readlines()
+        filtered_lines = [line.strip().split() for line in lines if re.match(
+                    "^[0-9][0-9]:[0-9][0-9]:[0-9][0-9]", line)]
+        cpu_list = []
+        usr_list = []
+        for columns in filtered_lines:
+            cpu_list.append(columns[2])
+            usr_list.append(columns[3])
+
+        if len(cpu_list) == on_cpu_count:
+            data = {
+                    'CPU': cpu_list,
+                    '%USER': usr_list,
+                    }
+            df = pd.DataFrame(data)
+            self.log.info("\n %s", df)
+            count = 0
+            for utilization in usr_list:
+                if int(float(utilization)) == int(self.capacity) or \
+                        int(float(utilization)) >= (int(self.capacity) - 2):
```

Comment:
what's the second condition is for?

Reply:
> what's the second condition is for?

I am attempting to address an issue where, during the execution of stress-ng, the reported CPU utilization by mpstat does not consistently reach 100%. Instead, it fluctuates between 98% and 100%. To ensure more accurate monitoring of CPU utilization, I am implementing a soft check.
`    int(float(utilization)) >= (int(self.capacity) - 2):`

Reply:
with `2` as offset, test failed few times.

I adjusted the offset to `5` and it worked.